File: MCPForUnity/Editor/Tools/Input/ManageInput.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using MCPForUnity.Editor.Helpers;
using UnityEditor;
using UnityEngine;
using UnityEngine.EventSystems;

namespace MCPForUnity.Editor.Tools.Input
{
    /// <summary>
    /// Simulates gameplay input during Play Mode (keyboard, mouse, touch, gamepad, UI).
    /// Requires Play Mode to be active for all actions except get_status.
    ///
    /// KEYBOARD:
    ///   - key_down: Press and hold a key
    ///   - key_up: Release a key
    ///   - key_press: Press + delay + release (async)
    ///
    /// MOUSE:
    ///   - mouse_move: Move mouse to screen coordinates [x, y]
    ///   - mouse_button_down: Press mouse button (0=left, 1=right, 2=middle)
    ///   - mouse_button_up: Release mouse button
    ///   - mouse_click: Click (down + delay + up, async)
    ///   - mouse_scroll: Scroll wheel
    ///
    /// TOUCH (New Input System only):
    ///   - touch: Simulate touch (phase: began/moved/ended/canceled)
    ///
    /// GAMEPAD (New Input System only):
    ///   - gamepad_button: Press/release gamepad button
    ///   - gamepad_axis: Set gamepad axis value
    ///
    /// UI (any input system):
    ///   - click_ui: Click a UI element by name or hierarchy path
    ///
    /// QUERY:
    ///   - get_input_state: Read current input device states
    ///   - get_status: Report capabilities and active input system
    ///
    /// SEQUENCES:
    ///   - send_sequence: Execute timed input sequence (returns job_id)
    ///   - get_sequence_status: Poll sequence job progress
    /// </summary>
    [McpForUnityTool("manage_input", AutoRegister = false, Group = "gameplay")]
    public static class ManageInput
    {
        private static readonly HashSet<string> PlayModeExemptActions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "get_status"
        };

        private static readonly HashSet<string> AllActions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "key_down", "key_up", "key_press",
            "mouse_move", "mouse_button_down", "mouse_button_up", "mouse_click", "mouse_scroll",
            "touch",
            "gamepad_button", "gamepad_axis",
            "click_ui",
            "get_input_state", "get_status",
            "send_sequence", "get_sequence_status"
        };

        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            var p = new ToolParams(@params);
            var actionResult = p.GetRequired("action");
            if (!actionResult.IsSuccess)
                return new ErrorResponse(actionResult.ErrorMessage);

            string action = actionResult.Value.ToLowerInvariant();

            if (!AllActions.Contains(action))
            {
                var suggestions = AllActions
                    .Where(a => a.StartsWith(action.Split('_')[0] + "_", StringComparison.OrdinalIgnoreCase))
                    .ToList();

                if (suggestions.Count > 0)
                    return new ErrorResponse($"Unknown action '{action}'. Similar actions: {string.Join(", ", suggestions)}");

                return new ErrorResponse(
                    $"Unknown action '{action}'. Valid actions: {string.Join(", ", AllActions.OrderBy(a => a))}");
            }

            if (!PlayModeExemptActions.Contains(action) && !EditorApplication.isPlaying)
            {
                return new ErrorResponse(
                    "Input simulation requires Play Mode. Use manage_editor action='play' first.");
            }

            try
            {
                return ExecuteAction(action, p, @params);
            }
            catch (Exception e)
            {
                return new ErrorResponse($"Input action '{action}' failed: {e.Message}");
            }
        }

        internal static object ExecuteAction(string action, ToolParams p, JObject @params)
        {
            switch (action)
            {
                // Keyboard
                case "key_down":
                    return HandleKeyDown(p);
                case "key_up":
                    return HandleKeyUp(p);
                case "key_press":
                    return HandleKeyPress(p);

                // Mouse
                case "mouse_move":
                    return HandleMouseMove(p);
                case "mouse_button_down":
                    return HandleMouseButtonDown(p);
                case "mouse_button_up":
                    return HandleMouseButtonUp(p);
                case "mouse_click":
                    return HandleMouseClick(p);
                case "mouse_scroll":
                    return HandleMouseScroll(p);

                // Touch
                case "touch":
                    return HandleTouch(p);

                // Gamepad
                case "gamepad_button":
                    return HandleGamepadButton(p);
                case "gamepad_axis":
                    return HandleGamepadAxis(p);

                // UI
                case "click_ui":
                    return UIInputSimulator.ClickUI(p);

                // Query
                case "get_input_state":
                    return InputSimulator.GetInputState();
                case "get_status":
                    return HandleGetStatus();

                // Sequences
                case "send_sequence":
                    return InputSequenceRunner.StartSequence(@params);
                case "get_sequence_status":
                    return InputSequenceRunner.GetStatus(p);

                default:
                    return new ErrorResponse($"Action '{action}' is recognized but not yet implemented.");
            }
        }

        // --- Keyboard ---

        private static object HandleKeyDown(ToolParams p)
        {
            var keyResult = p.GetRequired("key");
            if (!keyResult.IsSuccess)
                return new ErrorResponse(keyResult.ErrorMessage);

            if (InputSimulator.IsAvailable)
                return InputSimulator.KeyDown(keyResult.Value);

            return LegacyInputBridge.KeyDown(keyResult.Value);
        }

        private static object HandleKeyUp(ToolParams p)
        {
            var keyResult = p.GetRequired("key");
            if (!keyResult.IsSuccess)
                return new ErrorResponse(keyResult.ErrorMessage);

            if (InputSimulator.IsAvailable)
                return InputSimulator.KeyUp(keyResult.Value);

            return LegacyInputBridge.KeyUp(keyResult.Value);
        }

        private static object HandleKeyPress(ToolParams p)
        {
            var keyResult = p.GetRequired("key");
            if (!keyResult.IsSuccess)
                return new ErrorResponse(keyResult.ErrorMessage);

            float duration = p.GetFloat("duration") ?? 0.1f;

            if (InputSimulator.IsAvailable)
                return InputSimulator.KeyPress(keyResult.Value, duration);

            return LegacyInputBridge.KeyPress(keyResult.Value, duration);
        }

        // --- Mouse ---

        private static object HandleMouseMove(ToolParams p)
        {
            var posToken = p.GetRaw("position");
            if (posToken == null)
                return new ErrorResponse("Required parameter 'position' is missing. Provide [x, y] screen coordinates.");

            Vector2 pos;
            try
            {
                var arr = posToken.ToObject<float[]>();
                if (arr == null || arr.Length < 2)
                    return new ErrorResponse("'position' must be an array of [x, y].");
                pos = new Vector2(arr[0], arr[1]);
            }
            catch
            {
                return new ErrorResponse("'position' must be an array of [x, y] floats.");
            }

            if (InputSimulator.IsAvailable)
                return InputSimulator.MouseMove(pos);

            return LegacyInputBridge.MouseMove(pos);
        }

        private static object HandleMouseButtonDown(ToolParams p)
        {
            int button = p.GetInt("button") ?? 0;
            if (button < 0 || button > 2)
                return new ErrorResponse("'button' must be 0 (left), 1 (right), or 2 (middle).");

            if (InputSimulator.IsAvailable)
                return InputSimulator.MouseButtonDown(button);

            return LegacyInputBridge.MouseButtonDown(button);
        }

        private static object HandleMouseButtonUp(ToolParams p)
        {
            int button = p.GetInt("button") ?? 0;
            if (button < 0 || button > 2)
                return new ErrorResponse("'button' must be 0 (left), 1 (right), or 2 (middle).");

            if (InputSimulator.IsAvailable)
                return InputSimulator.MouseButtonUp(button);

            return LegacyInputBridge.MouseButtonUp(button);
        }

        private static object HandleMouseClick(ToolParams p)
        {
            int button = p.GetInt("button") ?? 0;
            float duration = p.GetFloat("duration") ?? 0.05f;

            var posToken = p.GetRaw("position");
            Vector2? pos = null;
            if (posToken != null)
            {
                try
                {
                    var arr = posToken.ToObject<float[]>();
                    if (arr != null && arr.Length >= 2)
                        pos = new Vector2(arr[0], arr[1]);
                }
                catch { /* ignore, will click at current position */ }
            }

            if (InputSimulator.IsAvailable)
                return InputSimulator.MouseClick(button, pos, duration);

            return LegacyInputBridge.MouseClick(button, pos, duration);
        }

        private static object HandleMouseScroll(ToolParams p)
        {
            float delta = p.GetFloat("scroll_delta") ?? p.GetFloat("scrollDelta") ?? 0f;
            if (Mathf.Approximately(delta, 0f))
                return new ErrorResponse("'scroll_delta' is required and must be non-zero.");

            if (InputSimulator.IsAvailable)
                return InputSimulator.MouseScroll(delta);

            return LegacyInputBridge.MouseScroll(delta);
        }

        // --- Touch ---

        private static object HandleTouch(ToolParams p)
        {
            if (!InputSimulator.IsAvailable)
                return new ErrorResponse("Touch simulation requires the New Input System package (com.unity.inputsystem).");

            var phaseStr = p.Get("phase") ?? "began";
            int fingerId = p.GetInt("finger_id") ?? p.GetInt("fingerId") ?? 0;

            var posToken = p.GetRaw("position");
            Vector2 pos = Vector2.zero;
            if (posToken != null)
            {
                try
                {
                    var arr = posToken.ToObject<float[]>();
                    if (arr != null && arr.Length >= 2)
                        pos = new Vector2(arr[0], arr[1]);
                }
                catch { /* default to zero */ }
            }

            return InputSimulator.Touch(fingerId, phaseStr, pos);
        }

        // --- Gamepad ---

        private static object HandleGamepadButton(ToolParams p)
        {
            if (!InputSimulator.IsAvailable)
                return new ErrorResponse("Gamepad simulation requires the New Input System package (com.unity.inputsystem).");

            var buttonResult = p.GetRequired("button_name");
            if (!buttonResult.IsSuccess)
                return new ErrorResponse(buttonResult.ErrorMessage);

            bool pressed = p.GetBool("pressed", true);

            return InputSimulator.GamepadButton(buttonResult.Value, pressed);
        }

        private static object HandleGamepadAxis(ToolParams p)
        {
            if (!InputSimulator.IsAvailable)
                return new ErrorResponse("Gamepad simulation requires the New Input System package (com.unity.inputsystem).");

            var axisResult = p.GetRequired("axis_name");
            if (!axisResult.IsSuccess)
                return new ErrorResponse(axisResult.ErrorMessage);

            var valueToken = p.GetRaw("axis_value") ?? p.GetRaw("axisValue");
            if (valueToken == null)
                return new ErrorResponse("Required parameter 'axis_value' is missing.");

            return InputSimulator.GamepadAxis(axisResult.Value, valueToken);
        }

        // --- Status ---

        private static object HandleGetStatus()
        {
            bool newIS = InputSimulator.IsAvailable;
            bool eventSystem = EventSystem.current != null;

            var supported = new List<string> { "get_status" };
            var limitations = new List<string>();

            if (EditorApplication.isPlaying)
            {
                if (newIS)
                {
                    supported.AddRange(new[]
                    {
                        "key_down", "key_up", "key_press",
                        "mouse_move", "mouse_button_down", "mouse_button_up", "mouse_click", "mouse_scroll",
                        "touch", "gamepad_button", "gamepad_axis",
                        "get_input_state", "send_sequence", "get_sequence_status"
                    });
                }
                else
                {
                    supported.AddRange(new[] { "key_down", "key_up", "key_press", "mouse_move",
                        "mouse_button_down", "mouse_button_up", "mouse_click", "mouse_scroll",
                        "send_sequence", "get_sequence_status" });
                    limitations.Add("Legacy Input Manager detected. Games using Input.GetKey()/GetAxis() may not receive simulated input. " +
                                    "Consider using the New Input System (com.unity.inputsystem) for full simulation support, " +
                                    "or use manage_components set_property to directly drive game logic.");
                    limitations.Add("Touch and gamepad simulation require New Input System package.");
                }

                if (eventSystem)
                    supported.Add("click_ui");
                else
                    limitations.Add("No EventSystem found in scene. click_ui action unavailable.");
            }
            else
            {
                limitations.Add("Not in Play Mode. All input actions except get_status require Play Mode.");
            }

            return new SuccessResponse("Input simulation status.", new
            {
                play_mode = EditorApplication.isPlaying,
                new_input_system = newIS,
                legacy_input = !newIS,
                event_system_present = eventSystem,
                bridge_active = LegacyInputBridge.IsBridgeActive,
                supported_actions = supported,
                limitations = limitations
            });
        }
    }
}

File: Server/src/services/tools/manage_input.py
from typing import Annotated, Any, Literal, Optional

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry

# All supported actions grouped by category
KEYBOARD_ACTIONS = ["key_down", "key_up", "key_press"]
MOUSE_ACTIONS = ["mouse_move", "mouse_button_down", "mouse_button_up", "mouse_click", "mouse_scroll"]
TOUCH_ACTIONS = ["touch"]
GAMEPAD_ACTIONS = ["gamepad_button", "gamepad_axis"]
UI_ACTIONS = ["click_ui"]
QUERY_ACTIONS = ["get_input_state", "get_status"]
SEQUENCE_ACTIONS = ["send_sequence", "get_sequence_status"]

ALL_ACTIONS = (
    KEYBOARD_ACTIONS + MOUSE_ACTIONS + TOUCH_ACTIONS +
    GAMEPAD_ACTIONS + UI_ACTIONS + QUERY_ACTIONS + SEQUENCE_ACTIONS
)


@mcp_for_unity_tool(
    group="gameplay",
    description=(
        "Simulates gameplay input during Play Mode (keyboard, mouse, touch, gamepad, UI). "
        "Requires Play Mode to be active (use manage_editor action='play' first). "
        "Best with New Input System package (com.unity.inputsystem); limited legacy Input support. "
        "Actions: key_down, key_up, key_press, mouse_move, mouse_button_down, mouse_button_up, "
        "mouse_click, mouse_scroll, touch, gamepad_button, gamepad_axis, click_ui, "
        "get_input_state, get_status, send_sequence, get_sequence_status."
    ),
    annotations=ToolAnnotations(
        title="Manage Input",
        destructiveHint=True,
    ),
)
async def manage_input(
    ctx: Context,
    action: Annotated[str, "Input action to perform."],
    key: Annotated[
        Optional[str],
        "Key name for keyboard actions (e.g., 'W', 'Space', 'LeftShift', 'Escape').",
    ] = None,
    button: Annotated[
        Optional[int],
        "Mouse button index: 0=left, 1=right, 2=middle.",
    ] = None,
    position: Annotated[
        Optional[list[float]],
        "Screen coordinates [x, y] for mouse_move, mouse_click, or touch.",
    ] = None,
    scroll_delta: Annotated[
        Optional[float],
        "Scroll wheel delta value for mouse_scroll.",
    ] = None,
    phase: Annotated[
        Optional[str],
        "Touch phase: 'began', 'moved', 'ended', 'canceled'.",
    ] = None,
    finger_id: Annotated[
        Optional[int],
        "Touch finger ID (0-9) for touch actions.",
    ] = None,
    button_name: Annotated[
        Optional[str],
        "Gamepad button name (e.g., 'south'/'a', 'north'/'y', 'start', 'leftShoulder').",
    ] = None,
    pressed: Annotated[
        Optional[bool],
        "Whether button is pressed (true) or released (false). Default true.",
    ] = None,
    axis_name: Annotated[
        Optional[str],
        "Gamepad axis name: 'leftStick', 'rightStick', 'leftTrigger', 'rightTrigger'.",
    ] = None,
    axis_value: Annotated[
        Optional[list[float] | float],
        "Axis value: [x, y] for sticks, float for triggers.",
    ] = None,
    element_name: Annotated[
        Optional[str],
        "UI element name or hierarchy path for click_ui (e.g., 'StartButton', 'Canvas/Panel/Button').",
    ] = None,
    duration: Annotated[
        Optional[float],
        "Duration in seconds for key_press/mouse_click (default 0.1s for keys, 0.05s for mouse).",
    ] = None,
    sequence: Annotated[
        Optional[list[dict]],
        "Array of steps for send_sequence. Each step: {action, params?, delay?}.",
    ] = None,
    job_id: Annotated[
        Optional[str],
        "Job ID for get_sequence_status polling.",
    ] = None,
    wait_timeout: Annotated[
        Optional[int],
        "Seconds to wait for sequence completion in get_sequence_status.",
    ] = None,
) -> dict[str, Any]:
    """Simulate gameplay input during Play Mode."""

    action_normalized = action.lower()

    if action_normalized not in ALL_ACTIONS:
        # Find closest matches by prefix
        prefix = action_normalized.split("_")[0] + "_" if "_" in action_normalized else ""
        available_by_prefix = {
            "key_": KEYBOARD_ACTIONS,
            "mouse_": MOUSE_ACTIONS,
            "touch": TOUCH_ACTIONS,
            "gamepad_": GAMEPAD_ACTIONS,
            "click_": UI_ACTIONS,
            "get_": QUERY_ACTIONS,
            "send_": SEQUENCE_ACTIONS,
        }
        suggestions = available_by_prefix.get(prefix, [])
        if suggestions:
            return {
                "success": False,
                "message": f"Unknown action '{action}'. Available {prefix}* actions: {', '.join(suggestions)}",
            }
        return {
            "success": False,
            "message": (
                f"Unknown action '{action}'. Valid actions: {', '.join(sorted(ALL_ACTIONS))}. "
                "Use action='get_status' to check available capabilities."
            ),
        }

    unity_instance = await get_unity_instance_from_context(ctx)

    # Build params dict, forwarding all non-None parameters
    params_dict: dict[str, Any] = {"action": action_normalized}

    if key is not None:
        params_dict["key"] = key
    if button is not None:
        params_dict["button"] = button
    if position is not None:
        params_dict["position"] = position
    if scroll_delta is not None:
        params_dict["scroll_delta"] = scroll_delta
    if phase is not None:
        params_dict["phase"] = phase
    if finger_id is not None:
        params_dict["finger_id"] = finger_id
    if button_name is not None:
        params_dict["button_name"] = button_name
    if pressed is not None:
        params_dict["pressed"] = pressed
    if axis_name is not None:
        params_dict["axis_name"] = axis_name
    if axis_value is not None:
        params_dict["axis_value"] = axis_value
    if element_name is not None:
        params_dict["element_name"] = element_name
    if duration is not None:
        params_dict["duration"] = duration
    if sequence is not None:
        params_dict["sequence"] = sequence
    if job_id is not None:
        params_dict["job_id"] = job_id

    params_dict = {k: v for k, v in params_dict.items() if v is not None}

    # For get_sequence_status with wait_timeout, implement server-side polling
    if action_normalized == "get_sequence_status" and wait_timeout and wait_timeout > 0:
        import asyncio
        import time

        deadline = time.monotonic() + wait_timeout
        poll_interval = 2.0

        while True:
            result = await send_with_unity_instance(
                async_send_command_with_retry,
                unity_instance,
                "manage_input",
                params_dict,
            )

            if not isinstance(result, dict):
                return {"success": False, "message": str(result)}

            # Check if terminal status
            data = result.get("data", {})
            status = data.get("status", "")
            if status in ("completed", "failed", "cancelled"):
                return result

            if time.monotonic() >= deadline:
                return result

            await asyncio.sleep(min(poll_interval, deadline - time.monotonic()))
    else:
        result = await send_with_unity_instance(
            async_send_command_with_retry,
            unity_instance,
            "manage_input",
            params_dict,
        )

        return result if isinstance(result, dict) else {"success": False, "message": str(result)}
